### PR TITLE
Added module support, and available Astilectron and Electron

### PR DESCRIPTION
**Pull request by alexj212**
This will allow go-astilectron-bundler to config the version of Electron and Astilectron that are to be downloaded and used for runtime.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,4 @@
 .DS_Store
 Thumbs.db
 .idea/
+go.sum
```
Comment (asticode) [inline .gitignore]: I'd rather not complexify the PR with adding module support. Can you remove this and make it in another PR?

File: Makefile [hunk 2]
```diff
@@ -1,2 +1,28 @@
-all:
-	go install ./astilectron-bundler
\ No newline at end of file
+export BIN_DIR=./bin
```
Comment (asticode) [inline Makefile]: I'd rather not complexify the PR with updating the Makefile. Can you remove this and make it in another PR?

File: go.mod [hunk 3]
```diff
@@ -0,0 +1,17 @@
+module github.com/asticode/go-astilectron-bundler
```
Comment (asticode) [inline go.mod]: I'd rather not complexify the PR with adding module support. Can you remove this and make it in another PR?

File: README.md [hunk 4]
```diff
@@ -27,12 +27,20 @@ Here's the basic configuration you'll usually need:
...
-  "icon_path_windows": "path/to/icon.ico"
+  "icon_path_windows": "path/to/icon.ico",
+  "version_electron": "4.0.1",
+  "version_astilectron": "0.33.0",
 }
 ```
 
 It will process the project located in the current directory and bundle it in the `output` dir for your os/arch.
 
+## Bundle Customizations
```
Comment (asticode) [inline README.md]: I'd rather rename the chapter to `Bundle for specific Astilectron and/or Electron versions` and remove the `app_name` line

File: astilectron-bundler/main.go [hunk 5]
```diff
@@ -3,6 +3,7 @@ package main
 import (
 	"encoding/json"
 	"flag"
+	"github.com/asticode/go-astilectron"
```
Comment (asticode) [inline astilectron-bundler/main.go]: Can you run `gofmt` ?

File: astilectron-bundler/main.go [hunk 6]
```diff
@@ -92,6 +93,20 @@ func main() {
 	}
 	c.LDFlags.Merge(astibundler.LDFlags(ldflags))
 
+	if c.VersionElectron == "" {
```
Comment (asticode) [inline astilectron-bundler/main.go]: Defaults shouldn't be set in the `main` package but in the lib instead. The `New()` function would be a perfect match

File: astilectron-bundler/main.go [hunk 7]
```diff
@@ -92,6 +93,20 @@ func main() {
...
+		c.VersionElectron = astilectron.DefaultVersionElectron
+		astilog.Debugf("Using VersionElectron - default: %v\n", c.VersionElectron)
+	} else {
+		astilog.Debugf("Using VersionElectron - override: %v\n", c.VersionElectron)
+	}
+
+
+	astilog.Debugf("astilectron.DefaultVersionElectron: %v\n", astilectron.DefaultVersionElectron)
```
Comment (asticode) [inline astilectron-bundler/main.go]: Can you remove thoses debug logs as well?

File: bundler.go [hunk 8]
```diff
@@ -115,6 +121,8 @@ type ConfigurationResourcesAdapter struct {
 type Bundler struct {
 	appName              string
 	bindPackage          string
+	versionElectron      string
```
Comment (asticode) [inline bundler.go]: Can you order attributes alphabetically?

File: bundler.go [hunk 9]
```diff
@@ -160,14 +168,16 @@ func absPath(configPath string, defaultPathFn func() (string, error)) (o string,
...
-		infoPlist:         c.InfoPlist,
+		appName:            c.AppName,
+		bindPackage:        c.Bind.Package,
+		Client:             &http.Client{},
+		environments:       c.Environments,
+		darwinAgentApp:     c.DarwinAgentApp,
+		resourcesAdapters:  c.ResourcesAdapters,
+		ldflags:            c.LDFlags,
+		infoPlist:          c.InfoPlist,
+		versionElectron:    c.VersionElectron,
```
Comment (asticode) [inline bundler.go]: Set both versions to defaults here. Then, at the bottom of this function, do something like this for each version:

```go
if c.VersionElectron != "" {
  b.versionElectron = c.VersionElectron
}
```

File: bundler.go [hunk 10]
```diff
@@ -314,6 +324,8 @@ func (b *Bundler) ClearCache() (err error) {
 func (b *Bundler) Bundle() (err error) {
 	// Create the output folder
 	astilog.Debugf("Creating %s", b.pathOutput)
+	astilog.Debugf("b.versionElectron %s", b.versionElectron)
```
Comment (asticode) [inline bundler.go]: Can you remove this debug log?

File: bundler.go [hunk 11]
```diff
@@ -483,8 +495,11 @@ func (b *Bundler) provisionVendor(oS, arch string) (err error) {
 		return
 	}
 
+
+	astilog.Debugf("provisionVendor b.versionElectron %s", b.versionElectron)
```
Comment (asticode) [inline bundler.go]: Can you remove this debug log?

File: bundler.go [hunk 12]
```diff
@@ -483,8 +495,11 @@ func (b *Bundler) provisionVendor(oS, arch string) (err error) {
...
+
 	// Provision electron
-	if err = b.provisionVendorElectron(oS, arch); err != nil {
+	if err = b.provisionVendorElectron(oS, arch, b.versionElectron); err != nil {
```
Comment (asticode) [inline bundler.go]: You shouldn't need to inject this parameter since `provisionVendorElectron` is a `Bundler` method and therefore has already the value of this parameter in its struct

File: bundler.go [hunk 13]
```diff
@@ -538,12 +554,17 @@ func (b *Bundler) provisionVendorAstilectron() (err error) {
...
 		}
 	}
-	return b.provisionVendorZip(astilectron.AstilectronDownloadSrc(), p, filepath.Join(b.pathVendor, zipNameAstilectron))
+	return b.provisionVendorZip(astilectron.AstilectronDownloadSrc(b.versionAstilectron), p, filepath.Join(b.pathVendor, zipNameAstilectron))
 }
 
 // provisionVendorElectron provisions the electron vendor zip file
-func (b *Bundler) provisionVendorElectron(oS, arch string) error {
-	return b.provisionVendorZip(astilectron.ElectronDownloadSrc(oS, arch), filepath.Join(b.pathCache, fmt.Sprintf("electron-%s-%s-%s.zip", oS, arch, astilectron.VersionElectron)), filepath.Join(b.pathVendor, zipNameElectron))
+func (b *Bundler) provisionVendorElectron(oS, arch, versionElectron string) error {
```
Comment (asticode) [inline bundler.go]: Remove the `versionElectron ` attribute

File: bundler.go [hunk 14]
```diff
@@ -538,12 +554,17 @@ func (b *Bundler) provisionVendorAstilectron() (err error) {
...
+	astilog.Debugf("provisionVendorElectron versionElectron %s", versionElectron)
```
Comment (asticode) [inline bundler.go]: Can you remove this debug log?

File: bundler.go [hunk 15]
```diff
@@ -538,12 +554,17 @@ func (b *Bundler) provisionVendorAstilectron() (err error) {
...
+
+	return b.provisionVendorZip(
+		astilectron.ElectronDownloadSrc(oS, arch, b.versionElectron),
+		filepath.Join(b.pathCache, fmt.Sprintf("electron-%s-%s-%s.zip", oS, arch, versionElectron)),
```
Comment (asticode) [inline bundler.go]: Use `b.versionElectron` instead of `versionElectron`

File: .gitignore (hunk 1 above)
Comment (alexj212) [reply .gitignore]: Removed module support, left for future PR

File: Makefile (hunk 2 above)
Comment (alexj212) [reply Makefile]: Removed module support and Makefile, left for future PR

File: README.md (hunk 4 above)
Comment (alexj212) [reply README.md]: updated README as per conversation

File: bundler.go (hunk 11 above)
Comment (alexj212) [reply bundler.go]: Removed debug messages

File: bundler.go (hunk 12 above)
Comment (alexj212) [reply bundler.go]: Fixed to use versions stored in Bundler.

File: bundler.go (hunk 13 above)
Comment (alexj212) [reply bundler.go]: Same as above

File: bundler.go (hunk 14 above)
Comment (alexj212) [reply bundler.go]: done

File: bundler.go (hunk 15 above)
Comment (alexj212) [reply bundler.go]: Set to use versions in Bundler

**Comment by alexj212** [comment]
Changes requested have been made

**Comment by asticode** [comment]
LGTM

Waiting for `go-astilectron` PR to be merged
